kernel/preemption/src/lib.rs

```diff
@@ -58,25 +42,19 @@ pub fn hold_preemption_no_timer_disable() -> PreemptionGuard {
 /// will be disabled upon a transition from preemption being enabled to being disabled.
 fn hold_preemption_internal<const DISABLE_TIMER: bool>() -> PreemptionGuard {
     let cpu_id = cpu::current_cpu();
+
     // Create an initial preemption guard such that we can call `CpuLocal::with_preempt()`,
     // but don't allow it to be dropped until we actually disable preemption below.
-    let mut guard_placeholder = NoDrop::new(PreemptionGuard {
+    PREEMPTION_COUNT.increment();
+    // Not a race condition because preemption is disabled.
+    let prev_val = PREEMPTION_COUNT.load() - 1;
```

Comment:
hmm, that's true in the simple case, but that only holds if we do not ever mess with preemption in an interrupt handler.

The increment and subsequent load are still not atomic, unlike the prior approach that used actual atomics (`fetch_add`). It's possible for an interupt to occur after the `increment` but before the `load`, and if that interrupt handler obtains or drops a `PreemptionGuard` and then returns, then the `prev_val` will be incorrect.

I assume there is no such single instruction that can increment a register _and_ return its prior value.... so without that, we have to disable interrupts around this sequence of two operations, right?

Reply:
Note: resolved via discord discussion, we'll use an atomic instruction `xadd` that returns the prior value instead of `inc`